### PR TITLE
Implemented M Coloring Problem

### PR DESCRIPTION
**Pull request by vedas-dixit**
[Open in Gitpod](https://gitpod.io/from-referrer/)&nbsp;[know more](https://www.gitpod.io/docs/pull-requests/)

### Describe your change:

- [x ] Add an algorithm?
- [ ] Fix a bug or typo in an existing algorithm?
- [ ] Documentation change?

### Checklist:

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Javascript/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized.
- [x] I know that pull requests will not be merged if they fail the automated tests.
- [x ] This PR only changes one algorithm file. To ease review, please open separate PRs for separate algorithms.
- [x ] All new JavaScript files are placed inside an existing directory.
- [x] All filenames should use the UpperCamelCase (PascalCase) style. There should be no spaces in filenames.
      **Example:**`UserProfile.js` is allowed but `userprofile.js`,`Userprofile.js`,`user-Profile.js`,`userProfile.js` are not
- [x] All new algorithms have a URL in their comments that points to Wikipedia or another similar explanation.
- [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: Backtracking/MColoringProblem.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
+/*
```
Comment (appgurueu) [inline Backtracking/MColoringProblem.js:1]: Use a proper doc comment instead of this.

File: Backtracking/MColoringProblem.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  Problem: M Coloring Problem
+  Given a graph represented as an adjacency matrix and a number M (number of colors), 
+  determine if the graph can be colored with up to M colors such that no two adjacent 
+  vertices share the same color.
+
+  What is the M Coloring Problem?
```
Comment (appgurueu) [inline Backtracking/MColoringProblem.js:7]: This is redundant with the description you have just given.

File: Backtracking/MColoringProblem.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  - This problem is about coloring a graph's vertices with at most M different colors 
+    such that no two adjacent vertices have the same color.
+
+  Example:
+  - Consider a triangle (3 nodes connected cyclically). We'd need 3 colors to color each vertex 
+    without adjacent vertices having the same color.
+
+  Solution:
```
Comment (appgurueu) [inline Backtracking/MColoringProblem.js:15]: These should be comments on the implementation, on the relevant parts.

File: Backtracking/MColoringProblem.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  - We will be using backtracking to solve this question.
+  - Color a vertex, then move to an adjacent vertex and try all colors. If a color is valid, 
+    continue to the next vertex, else backtrack.
+*/
+
+class MColoring {
```
Comment (appgurueu) [inline Backtracking/MColoringProblem.js:21]: There is no need for a class here. This should effectively be a single `function` with parameters `graph` and `m` which returns the colors, or `null` / `undefined` if there is no valid m-coloring. Note that using a function does not prevent you from using and defining "helper functions" *inside the function*.

**Review by appgurueu — CHANGES_REQUESTED** [review]
It also seems to me that an adjacency list representation would be more suitable?

**Comment by vedas-dixit** [comment]
Thank you @appgurueu for the detailed review. I appreciate the feedback and suggestions. I'll work on implementing the changes you recommended:

I'll consider switching to an adjacency list representation for better efficiency.
I'll refine the comments, utilize proper JS Doc comments, and ensure that explanations are placed next to the relevant code.
I'll revisit the class-based approach and try to refactor it to a functional one as you suggested.
I'll push these changes soon. Thanks again for the insights!

**Review by appgurueu — APPROVED** [review]

**Review by raklaptudirm — APPROVED** [review]

**Pull request merged**